Source/JavaScriptCore/runtime/SetPrototype.cpp

```diff
@@ -175,6 +181,175 @@ JSC_DEFINE_HOST_FUNCTION(setProtoFuncSize, (JSGlobalObject* globalObject, CallFr
+
+    ASSERT(otherValue.isObject());
+    JSObject* otherObject = asObject(otherValue);
+
+    JSValue has = otherObject->get(globalObject, vm.propertyNames->has);
+    RETURN_IF_EXCEPTION(scope, { });
+    if (!has.isCallable()) [[unlikely]]
+        return throwVMTypeError(globalObject, scope, "Set.prototype.intersection expects other.has to be callable");
+
+    JSValue keys = otherObject->get(globalObject, vm.propertyNames->keys);
+    RETURN_IF_EXCEPTION(scope, { });
+    if (!keys.isCallable()) [[unlikely]]
+        return throwVMTypeError(globalObject, scope, "Set.prototype.intersection expects other.keys to be callable");
+
+    JSSet* result = JSSet::create(vm, globalObject->setStructure());
+    if (thisSet->size() <= size) {
+        JSCell* storageCell = thisSet->storageOrSentinel(vm);
+        if (storageCell == vm.orderedHashTableSentinel())
+            return JSValue::encode(result);
+
+        JSSet::Storage& storage = *jsCast<JSSet::Storage*>(storageCell);
+        JSSet::Helper::Entry entry = 0;
+        CallData hasCallData = JSC::getCallData(has);
+
+        std::optional<CachedCall> cachedHasCall;
+        if (hasCallData.type == CallData::Type::JS) [[likely]] {
+            cachedHasCall.emplace(globalObject, jsCast<JSFunction*>(has), 1);
+            RETURN_IF_EXCEPTION(scope, { });
+        }
+
+        while (true) {
+            storageCell = JSSet::Helper::nextAndUpdateIterationEntry(vm, storage, entry);
+            if (storageCell == vm.orderedHashTableSentinel())
+                break;
+
+            storage = *jsCast<JSSet::Storage*>(storageCell);
+            entry = JSSet::Helper::iterationEntry(storage) + 1;
+            JSValue entryKey = JSSet::Helper::getIterationEntryKey(storage);
+
+            JSValue hasResult;
+            if (cachedHasCall) [[likely]] {
+                hasResult = cachedHasCall->callWithArguments(globalObject, otherValue, entryKey);
+                RETURN_IF_EXCEPTION(scope, { });
+            } else {
+                MarkedArgumentBuffer args;
+                args.append(entryKey);
+                ASSERT(!args.hasOverflowed());
+                hasResult = call(globalObject, has, hasCallData, otherValue, args);
+                RETURN_IF_EXCEPTION(scope, { });
+            }
+
+            bool hasResultBool = hasResult.toBoolean(globalObject);
+            RETURN_IF_EXCEPTION(scope, { });
+            if (hasResultBool) {
+                result->add(globalObject, entryKey);
+                RETURN_IF_EXCEPTION(scope, { });
+            }
+        }
+    } else {
+        CallData keysCallData = JSC::getCallData(keys);
+        JSValue iterator;
+        if (keysCallData.type == CallData::Type::JS) [[likely]] {
+            CachedCall cachedCall(globalObject, jsCast<JSFunction*>(keys), 0);
+            RETURN_IF_EXCEPTION(scope, { });
+            iterator = cachedCall.callWithArguments(globalObject, otherValue);
+            RETURN_IF_EXCEPTION(scope, { });
```

Comment:
If we do not call it multiple times, we should just use `call` (not CachedCall).